Modular CLI

LokiLuciferase opened the pull request:
Making recently introduced functionality (XGB, parameter search...) available via the `argparse` CLI interface would require most subcommands to be able to handle differing (but largely shared among subcommands) sets of options.
This would make the interface either excessively WET, or necessitate high number of confusing conditional statements.

To make our CLI more extendable, modular and testable, I have re-implemented it using the [click](https://www.github.com/pallets/click) framework. 

Looking forward to the package name change, the new CLI is (for now) invoked with `trex`.
The following commands are available:
- `predict`: equivalent to previous command `predict`
- `get-weights`: equivalent to previous command `weights`
- `train`: roughly equivalent to previous command, but takes as additional subcommand the type of classifier to train (`svm` or `xgb`). Allows passing of a training parameters file in JSON format using the flag `--params_file`.
- `cv`: roughly equivalent to previous command `crossvalidate`, same extras as `train`. Allows finding and saving of optimal training parameters using the options `--optimize` and `--optimize_out`. Functionality as previous command `logo` can be obtained with options `--groups` and `--rank`.
- `cccv`: roughly equivalent to previous command, same extras as `train`.
- `plot`: takes currently only the subcommand `cccv` which allows plotting of one or multiple CCCV results files in PhenDB format.

Click includes facilities for testing CLI components but as of now, the CLI is not tested (and is excluded from codecov).

Other minor changes:
- moved serialization functionality from utils to io
- bumped version
- added totally sick ascii art

LokiLuciferase committed:
Create generic, extendable CLI with Click.

LokiLuciferase committed:
Finalize click CLI and invert verb and object; move clf serialization to io.

LokiLuciferase committed:
attach parameter search to CLI and implement loading and saving of found parameters to JSON.

LokiLuciferase committed:
add cccv plotting facility and remove defunct original entry point

LokiLuciferase committed:
fix missing verbosity

VarIr commented on pica/cli/generic_func.py:
```diff
@@ -0,0 +1,110 @@
+               optimize_out=None, groups=None, rank=None, out=None, n_features=None, params_file=None,
+               *args, **kwargs):
+    """
+    Generic function for model CV. Optionally, perform parameter search
```
I'd rather have the first sentence similar to the second one, e.g. "Estimate model performance by cross-validation" (if that is what the function does).

VarIr commented on pica/cli/generic_func.py:
```diff
@@ -0,0 +1,110 @@
+        loaded_params = load_params_file(params_file)
+        logger.info(f'Parameters loaded from file:')
+        logger.info('\n' + pformat(loaded_params))
+        kwargs = {**kwargs, **loaded_params}  # TODO: should loaded params have precendence?
```
It seems the loaded params have precedence here (as in: if there are identical keys, `loaded_params` will overwrite those from `kwargs`), right?

LokiLuciferase replied:
yes, correct.

VarIr commented on pica/cli/generic_func.py:
```diff
@@ -0,0 +1,110 @@
+
+def generic_train(type, genotype, phenotype, verb, weights, out,
+                  n_features=None, params_file=None, *args, **kwargs):
+    """Generic function for model training."""
```
As in the other `generic_` functions, I'd prefer it like "Train some model"

VarIr commented on pica/cli/generic_func.py:
```diff
@@ -0,0 +1,110 @@
+def generic_cccv(type, genotype, phenotype, folds, replicates, threads, comple_steps, conta_steps,
+                 verb, groups=None, rank=None, optimize=False, out=None, n_features=None, params_file=None,
+                 *args, **kwargs):
+    """Generic function for model CCCV."""
```
The same here. Though if you prefer it like it is now, that's also fine. Should at least be consistent style (which it is).

Also, please don't use the abbrev in the docstring, so that we can immediately what CCCV stands for in this place.

VarIr commented on pica/cli/main.py:
```diff
@@ -0,0 +1,33 @@
+
+def main():
+    print("""
+    ██████╗ ██╗  ██╗███████╗███╗   ██╗ ██████╗       ████████╗██████╗ ███████╗██╗  ██╗
```
Don't know why, but on github this is slightly distorted, while it looks fine if I copy it into gedit.
Otherwise, +1 for ANSI shadow :)

LokiLuciferase replied:
think we can all agree this is the most significant improvement to trex since its inception.

VarIr commented on pica/cli/plot.py:
```diff
@@ -0,0 +1,29 @@
+@click.group()
+def plot():
+    """
+    Perform plotting of results. Conditions are named after CCCV input file names.
```
Perhaps simply: "Plot results. [...]"

VarIr commented on pica/cli/predict.py:
```diff
@@ -0,0 +1,23 @@
+    preds, probas = model.predict(X=gr)
+    translate_output = {trait_id: trait_sign for trait_sign, trait_id in
+                        DEFAULT_TRAIT_SIGN_MAPPING.items()}
+    sys.stdout.write("Identifier\tTrait present\tConfidence\n")
```
Why not use print here?

LokiLuciferase replied:
mumble mumble circumvent stdout buffer mumble mumble

...no idea in all honesty. I copied the code from the old predict function :)

VarIr commented on pica/io/flat.py:
```diff
@@ -115,6 +115,40 @@ def load_groups_file(input_file: str, selected_rank: str = None) -> List[GroupRe
+    :param params: A dictionary of training parameters.
+    :return: A dictionary of training parameters.
+    """
+    class NpEncoder(json.JSONEncoder):
```
`NumpyEncoder` would be more articulate (while slightly longer)

VarIr commented on pica/ml/classifiers/xgbm.py:
```diff
@@ -32,9 +32,9 @@ def __init__(self, max_depth: int = 4, learning_rate: float = 0.05,
             'subsample': np.arange(0.2, 1., 0.1).round(2),
             'colsample_bytree': np.arange(0.2, 1., 0.1).round(2),
-            'min_child_weight': np.arange(1, 20),
+            'min_child_weight': np.arange(1, 20).astype(int),
```
Why explicitly cast to int? Should be integer anyway.

VarIr commented on pica/ml/classifiers/xgbm.py:
```diff
@@ -32,9 +32,9 @@ def __init__(self, max_depth: int = 4, learning_rate: float = 0.05,
             'colsample_bytree': np.arange(0.2, 1., 0.1).round(2),
-            'min_child_weight': np.arange(1, 20),
+            'min_child_weight': np.arange(1, 20).astype(int),
             'gamma': [0, 0.2, 0.5, 1, 5, 10],
```
Here we have a mixture of Python lists and Numpy arrays. This does not matter as long as we only ever iterate over these structures. I'm only slightly worried that we at some point may do `gamma * 2` and obtain something rather surprising.

VarIr commented on tests/test_data/Sulfate_reducer_xgb_cccv.json:
```diff
@@ -0,0 +1,728 @@
+		"contamination": 1.0,
+		"completeness": 1.0
+	}
+]
```
I'm not so familiar with JSON, but POSIX standard would be to have a newline at EOF. I assume the JSON files are read/written by a library that just does not do that? Then it would probably not matter.

LokiLuciferase replied:
yeah, written and read with `json`. But inserting a newline won't hurt, you are right.

VarIr commented on tests/test_data/svm_params.json:
```diff
@@ -0,0 +1,6 @@
+  "penalty": "l2",
+  "C": 0.00057362,
+  "max_iter": 174
+}
```
Again, newline at EOF?

VarIr commented on tests/test_data/xgb_params.json:
```diff
@@ -0,0 +1,11 @@
+  "scale_pos_weight": 3,
+  "learning_rate": 0.1,
+  "eval_metric": "auc"
+}
```
Again, newline at EOF?

VarIr commented on tests/test_data/Sulfate_reducer_svm_cccv.json:
```diff
@@ -0,0 +1,728 @@
+		"contamination": 1.0,
+		"completeness": 1.0
+	}
+]
```
Again, newline at EOF?

VarIr reviewed (commented):
I did not know about click, but seems like a really cool package!
As usual, I have a few stylistic remarks, but in general this LGTM.
However, I just had a look at the code. Could you perhaps try to actually run the new cli, @phyden?

VarIr commented:
Btw, while high code cov is a nice thing to have, I would not enforce it. Even 100% is no guarantee for error-free code, and code with incomplete coverage can be highly reliable (e.g. SciPy has 87% cov for Python code and only 45% for complied code). Still, some mechanism to ensure a functioning cli is a must at least for new releases on PyPI etc. Could be a handful of manual checks though.

VarIr commented:
We have quite a number of flake8 warnings. I'll fix them (or set flake8 to ignore, in case of whitespace before a colon). Can I push to your feature branch? Or do I have to create a PR onto your PR?

LokiLuciferase commented:
By all means, please do. Also feel free to implement any changes discussed if you have time, I probably won't until weekend at earliest.  😁  Do you have push access to my fork?

VarIr commented:
Will try! I also don't have time actually, so I cannot guarantee for anything more.

LokiLuciferase committed:
address comments by @VarIr re PR#11

LokiLuciferase committed:
fix flake8 warnings

LokiLuciferase committed:
only print ASCII art when invoking `trex -h`

LokiLuciferase commented:
I found some time to apply the changes, thanks @VarIr ! By chance I found an issue with the CLI, guess it's really a good idea to test it through manually before merging and then write some tests soon.

The error: I forgot that some of the commands rely on a clean stdout stream, but the way I added my super rad ASCII art polluted it 🤷‍♂

VarIr commented on pica/cli/cccv.py:
```diff
@@ -8,6 +8,7 @@
 
+
 @click.group()
 def cccv():
```
When I run `trex -h`, the description of CCCV is truncated like "Perform nested, stratified crossvalidation over a range of..." with the rest missing.

I think we should have a short descprition for `trex -h`, like "Simulated completeness/contamination crossvalidation". The rest does appear when calling `trex cccv -h`.

For example:
``` python3
@click.group()
def cccv():
    """
    Simulated completeness/contamination crossvalidation.

    Perform nested, stratified crossvalidation over a range of simulated completeness
    and contamination values for the training data.
    Optionally, save returned scores for each grid point.
    """
    pass
```

VarIr reviewed (commented):
Thanks for the quick action! Just one more comment on the help texts of `click`, otherwise LGTM.

LokiLuciferase committed:
add short help strings to represent subcommands in main CLI interface

LokiLuciferase committed:
add repr methods for dataclasses to prevent printing of features

LokiLuciferase committed:
package name change to `phenotrex` and more explicit module privacy

LokiLuciferase committed:
guard against parallelizing both internally and by subprocessing which leads to BrokenProcessPool when using XGB. If both classifier-internal n_jobs and method n_jobs (realized via subprocessing) is not 1, classifier n_jobs takes precedence.

LokiLuciferase committed:
- fix cv verbosity via CLI
- make output file required for cccv
- fix pickled classifier (allow loading of old classifiers made with `pica`)

LokiLuciferase committed:
attach selection of n models to evaluate to cv CLI; better logging; rounded output probas for predict
